compiler/rustc_typeck/src/check/fn_ctxt/checks.rs

```diff
@@ -291,7 +317,42 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             } else {
                 false
             };
-            param_count_error(expected_arg_count, supplied_arg_count, "E0061", false, sugg_unit);
+
+            // are we passing elements of a tuple without the tuple parentheses?
+            let chosen_arg_tys = match &expected_arg_tys[..] {
+                [] => fn_inputs,
+                args @ [..] => args,
+            };
+
+            let sugg_tuple_wrap_args = chosen_arg_tys
+                .get(0)
+                .cloned()
+                .map(|arg_ty| self.resolve_vars_if_possible(arg_ty))
+                .and_then(|arg_ty| match arg_ty.kind() {
+                    ty::Tuple(tup_elems) => Some(tup_elems),
+                    _ => None,
+                })
+                .map(|tup_elems| tup_elems.len() == supplied_arg_count && chosen_arg_tys.len() == 1)
+                .and_then(|potential_tuple| {
+                    if potential_tuple {
+                        match args {
+                            [first, .., last] => Some((first, Some(last))),
+                            [single] => Some((single, None)),
```

Comment:
I wonder if it could be clearer to use a custom enum here? Something like

```rust
enum TupleAsFnArg<'hir> {
    Single(&'hir hir::Expr<'hir>),
    Multi { first: &'hir hir::Expr<'hir>, last: &'hir hir::Expr<'hir> }
}
```

Reply:
Yeah I agree - that's much nicer. I wasn't keen on the long type declaration on the closure argument either.